### PR TITLE
Add interhand3d detector, head and config. 

### PR DESCRIPTION
**Pull request by zengwang430521**
1. Add interhand3d detector, head and config. 
2. Fix some bugs in multilabel_classification_head and interhand3d_dataset.

### CODE REVIEW

File: mmpose/datasets/pipelines/pose3d_transform.py [hunk 1]
```diff
@@ -325,34 +327,60 @@ def __call__(self, results):
...
+        yy = yy.round().clip(0, H - 1)
+        zz = zz.round().clip(0, D - 1)
+
+        # compute the target value near joints
+        local_target = \
+            np.exp(-((xx - mu_x)**2 + (yy - mu_y)**2 + (zz - mu_z)**2) /
+                   (2 * self.sigma**2))
+
+        # put the local target value to the full target heatmap
+        tmp_size = xx.shape[1]
```
Comment (jin-s13) [inline mmpose/datasets/pipelines/pose3d_transform.py:375]: `tmp_size` is not a good name.

Comment (zengwang430521) [reply mmpose/datasets/pipelines/pose3d_transform.py:375]: I will change it to local_size

**Comment by jin-s13** [comment]
Please improve the unittest coverage if possible.

File: mmpose/datasets/pipelines/pose3d_transform.py [hunk 2]
```diff
@@ -325,34 +327,60 @@ def __call__(self, results):
...
+            np.exp(-((xx - mu_x)**2 + (yy - mu_y)**2 + (zz - mu_z)**2) /
+                   (2 * self.sigma**2))
+
+        # put the local target value to the full target heatmap
+        local_size = xx.shape[1]
+        idx_joints = np.tile(
+            np.arange(num_joints)[:, None, None, None],
+            [1, local_size, local_size, local_size])
+        idx = np.stack([idx_joints, zz, yy, xx],
+                       axis=-1).astype(np.long).reshape(-1, 4)
```
Comment (jin-s13) [inline mmpose/datasets/pipelines/pose3d_transform.py:380]: Do we really need np.long here?

Comment (zengwang430521) [reply mmpose/datasets/pipelines/pose3d_transform.py:380]: Yes, because zz, yy and xx are np.float.

**Comment by jin-s13** [comment]
CI fails.

File: mmpose/models/detectors/interhand_3d.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
+from ..registry import POSENETS
+from .top_down import TopDown
+
+
+@POSENETS.register_module()
+class Interhand3D(TopDown):
```
Comment (liuxin9608) [inline mmpose/models/detectors/interhand_3d.py:6]: remove neck

File: mmpose/core/evaluation/top_down_eval.py [hunk 4]
```diff
@@ -91,6 +91,43 @@ def _get_max_preds(heatmaps):
...
+    idx = np.argmax(heatmaps_reshaped, 2).reshape((N, K, 1))
+    maxvals = np.amax(heatmaps_reshaped, 2).reshape((N, K, 1))
+
+    preds = np.zeros((N, K, 3), dtype=np.float32)
+    _idx = idx[..., 0]
+    preds[:, :, 2], _idx = _idx // (H * W), _idx % (H * W)
+    preds[:, :, 1], _idx = _idx // W, _idx % W
+    preds[:, :, 0] = _idx
+
+    preds = np.where(np.tile(maxvals, (1, 1, 3)) > 0.0, preds, -1)
```
Comment (innerlee) [inline mmpose/core/evaluation/top_down_eval.py:127]: I guess the `np.tile` can be removed.

File: mmpose/core/evaluation/top_down_eval.py [hunk 5]
```diff
@@ -91,6 +91,43 @@ def _get_max_preds(heatmaps):
     return preds, maxvals
 
 
+def get_max_preds_3d(heatmaps):
```
Comment (innerlee) [inline mmpose/core/evaluation/top_down_eval.py:94]: I saw that `_get_max_preds` is not public. Will `get_max_preds_3d` be used outside?

File: mmpose/core/evaluation/top_down_eval.py [hunk 6]
```diff
@@ -574,3 +611,13 @@ def keypoints_from_heatmaps(heatmaps,
         maxvals = maxvals / 255.0 + 0.5
 
     return preds, maxvals
+
+
+def multilabel_classification_accuracy(pred, gt, mask, thr=0.5):
```
Comment (innerlee) [inline mmpose/core/evaluation/top_down_eval.py:616]: docstring

File: mmpose/core/evaluation/top_down_eval.py [hunk 7]
```diff
@@ -574,3 +611,13 @@ def keypoints_from_heatmaps(heatmaps,
...
+    valid = (mask > 0).min(axis=1) if mask.ndim == 2 else (mask > 0)
+    pred, gt = pred[valid], gt[valid]
+    if pred.shape[0] == 0:
+        acc = 0  # when no sample is with gt labels, set acc to 0.
+    else:
+        acc = (((pred - thr) * (gt - thr)).min(axis=1) > 0).mean()
```
Comment (innerlee) [inline mmpose/core/evaluation/top_down_eval.py:622]: What's the meaning of the axis 1?

File: mmpose/models/keypoint_heads/interhand_3d_head.py [hunk 8]
```diff
@@ -0,0 +1,529 @@
...
+        hidden_dims (list|tuple): Number of hidden dimension of FC layers.
+    """
+
+    def __init__(self, in_channels=2048, num_labels=2, hidden_dims=(512, )):
+        super().__init__()
+
+        self.in_channels = in_channels
+        self.num_labesl = num_labels
+
+        feature_dims = [in_channels] + \
```
Comment (innerlee) [inline mmpose/models/keypoint_heads/interhand_3d_head.py:257]: There is a shorter way

```python
[in_channels, *hidden_dims, num_labels]
```

**Comment by innerlee** [comment]
LGTM

File: mmpose/core/evaluation/top_down_eval.py (hunk 4 above)
Comment (zengwang430521) [reply mmpose/core/evaluation/top_down_eval.py:127]: Yes, I have removed it.

File: mmpose/core/evaluation/top_down_eval.py (hunk 5 above)
Comment (zengwang430521) [reply mmpose/core/evaluation/top_down_eval.py:94]: I have modify the codes and only use keypoints_from_heatmaps3d outside.

File: mmpose/core/evaluation/top_down_eval.py (hunk 6 above)
Comment (zengwang430521) [reply mmpose/core/evaluation/top_down_eval.py:616]: OK, added.

File: mmpose/core/evaluation/top_down_eval.py (hunk 7 above)
Comment (zengwang430521) [reply mmpose/core/evaluation/top_down_eval.py:622]: The classification of a sample is regarded as correct only if it's correct for all labels. 
I changed it to a way easier to understand.

File: mmpose/models/keypoint_heads/interhand_3d_head.py (hunk 8 above)
Comment (zengwang430521) [reply mmpose/models/keypoint_heads/interhand_3d_head.py:257]: OK, changed.

File: mmpose/core/evaluation/top_down_eval.py (hunk 5 above)
Comment (innerlee) [reply mmpose/core/evaluation/top_down_eval.py]: If `get_max_preds_3d` is not designed to be used outside, then we can rename it to `_get_max_preds_3d`.

Comment (innerlee) [reply mmpose/core/evaluation/top_down_eval.py]: Sorry looked at an old diff. Ignore the above comment :)

File: mmpose/models/keypoint_heads/interhand_3d_head.py [hunk 9]
```diff
@@ -0,0 +1,521 @@
...
+        hidden_dims (list|tuple): Number of feature dimension of FC layers.
+    """
+
+    def __init__(self, in_channels=2048, heatmap_size=64, hidden_dims=(512, )):
+        super().__init__()
+
+        self.in_channels = in_channels
+        self.heatmap_size = heatmap_size
+
+        feature_dims = [in_channels] + \
```
Comment (innerlee) [inline mmpose/models/keypoint_heads/interhand_3d_head.py:205]: here

Comment (zengwang430521) [reply mmpose/models/keypoint_heads/interhand_3d_head.py]: OK, fixed.

File: README_CN.md [hunk 10]
```diff
@@ -81,6 +81,7 @@ MMPose 是一款基于 PyTorch 的姿态分析的开源工具箱，是 [OpenMMLa
 - [x] [RSN](configs/top_down/rsn/README.md) (ECCV'2020)
 - [x] [HMR](configs/mesh/hmr/README.md) (CVPR'2018)
 - [x] [Simple 3D Baseline](configs/body3d/simple_baseline/README.md) (ICCV'2017)
+- [x] [InterNet](configs/hand3d/InterNet/README.md) (ArXiv'2020)
```
Comment (ly015) [inline README_CN.md:84]: ECCV2020?

File: configs/hand3d/InterNet/README.md [hunk 11]
```diff
@@ -0,0 +1,26 @@
+# InterHand2.6M: A Dataset and Baseline for 3D Interacting Hand Pose Estimation from a Single RGB Image
+
+## Introduction
+
+<!-- [ALGORITHM] -->
+
+```bibtex
```
Comment (ly015) [inline configs/hand3d/InterNet/README.md:7]: ECCV 2020?

File: mmpose/core/evaluation/top_down_eval.py [hunk 12]
```diff
@@ -91,6 +91,43 @@ def _get_max_preds(heatmaps):
...
+    N, K, D, H, W = heatmaps.shape
+    heatmaps_reshaped = heatmaps.reshape((N, K, -1))
+    idx = np.argmax(heatmaps_reshaped, 2).reshape((N, K, 1))
+    maxvals = np.amax(heatmaps_reshaped, 2).reshape((N, K, 1))
+
+    preds = np.zeros((N, K, 3), dtype=np.float32)
+    _idx = idx[..., 0]
+    preds[:, :, 2], _idx = _idx // (H * W), _idx % (H * W)
+    preds[:, :, 1], _idx = _idx // W, _idx % W
+    preds[:, :, 0] = _idx
```
Comment (ly015) [inline mmpose/core/evaluation/top_down_eval.py:125]: Do you think the following form would be more clear to users? Just a suggestion.

preds[...,2] = _idx // (H*W)
preds[...,1] = (_idx // W) % H
preds[...,0] = _idx % W

File: mmpose/models/detectors/interhand_3d.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+              Otherwise, return predicted poses, boxes, image paths, heatmaps,
+              relative hand root depth and hand type.
+        """
+        if return_loss:
+            return self.forward_train(img, target, target_weight, img_metas,
+                                      **kwargs)
+        return self.forward_test(img, img_metas, **kwargs)
+
+    def forward_test(self, img, img_metas, **kwargs):
+        """Defines the computation performed at every call when testing."""
```
Comment (ly015) [inline mmpose/models/detectors/interhand_3d.py:67]: It seems identical to TopDown's forward_test except for return_heatmap==False. Would it be better to use super().forward_test(..., return_heatmap=False, **kwargs)?

Comment (zengwang430521) [reply mmpose/models/detectors/interhand_3d.py:67]: https://github.com/open-mmlab/mmpose/blob/fd7eb677b7dd5f1c6751a9e6b4f12eb8e8f05bdc/mmpose/models/detectors/interhand_3d.py#L88-L89

The main difference is in line 88-89. The output of interhand_3d detector is a list.

File: README_CN.md (hunk 10 above)
Comment (zengwang430521) [reply README_CN.md]: Fixed.

File: configs/hand3d/InterNet/README.md (hunk 11 above)
Comment (zengwang430521) [reply configs/hand3d/InterNet/README.md:7]: Fixed.

File: mmpose/core/evaluation/top_down_eval.py (hunk 12 above)
Comment (zengwang430521) [reply mmpose/core/evaluation/top_down_eval.py]: OK, modified.

**Pull request merged**